File: Ladder02/Java_Code/MedianOfTwoSortedArrays.java
import java.util.*;

class MedianOfTwoSortedArrays {
    /**
     * @param A: An integer array.
     * @param B: An integer array.
     * @return: a double whose format is *.5 or *.0
     */
    public double findMedianSortedArrays(int[] A, int[] B) {
        int length = A.length + B.length;
        if (length % 2 == 0) {
            return (findKth(A, B, length / 2) + findKth(A, B, length / 2 + 1)) / 2.0;
        } else {
            return findKth(A, B, (length + 1) / 2);
        }
    }

    private int findKth(int[] A, int[] B, int k) {
        if (A.length > B.length) {
            return findKth(B, A, k);
        }
        if (A.length == 0) {
            return B[k - 1];
        }
        if (k == 1) {
            return Math.min(A[0], B[0]);
        }
        int aMid = Math.min(k / 2, A.length);
        int bMid = k - aMid;
        if (A[aMid - 1] < B[bMid - 1]) {
            return findKth(Arrays.copyOfRange(A, aMid, A.length), B, k - aMid);
        } else {
            return findKth(A, Arrays.copyOfRange(B, bMid, B.length), k - bMid);
        }
    }

    public static void main(String[] args) {
        MedianOfTwoSortedArrays test = new MedianOfTwoSortedArrays();
        int[] A = {1,2,3};
        int[] B = {4,5};
        System.out.println(test.findMedianSortedArrays(A, B));
    }
}

File: Ladder02/Python_code/median_of_two_sorted_arrays.py
class Solution:
    """
    @param A: An integer array.
    @param B: An integer array.
    @return: a double whose format is *.5 or *.0
    """
    def findMedianSortedArrays(self, A, B):
        length = len(A) + len(B)
        if length % 2 == 1:
            return self.findKth(A, B, length / 2 + 1)
        else:
            return (self.findKth(A, B, length / 2) +\
                    self.findKth(A, B, length / 2 + 1)) / 2.0

    def findKth(self, A, B, k):
        if len(A) > len(B):
            return self.findKth(B, A, k)
        else:
            if len(A) == 0:
                return B[k - 1]
            elif k == 1:
                return min(A[0], B[0])
            else:
                A_mid = min(k / 2, len(A))
                B_mid = k - A_mid
                if A[A_mid - 1] > B[B_mid - 1]:
                    return self.findKth(A, B[B_mid:], k - B_mid)
                else:
                    return self.findKth(A[A_mid:], B, k - A_mid)